### PR TITLE
Diff cocina from purl to determine what to shelve.

### LINKED ISSUE #4982: Use cocina structural diff to determine files to send to purl-fetcher
Currently, the files to send to purl-fetcher are determined based on the existing content diffing strategy: https://github.com/sul-dlss/dor-services-app/blob/main/app/services/shelving_service.rb#L50

For this ticket, implement a new strategy that compares the current cocina structural against the cocina structural retrieved from purl (https://purl.stanford.edu/<druid>.json). For any file which is (1) shelved and (2) not in purl cocina or in purl cocina but has different checksums, send to purl-fetcher.

### PR DESCRIPTION
**Pull request by justinlittman**
closes #4982

## Why was this change made? 🤔
Less janky.

## How was this change tested? 🤨

⚡ ⚠ If this change has cross service impact, including data writes to shared file systems, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

Unit, integration tests.

### CODE REVIEW

File: app/services/digital_stacks_differ.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    @cocina_object_file_map ||= file_map_for(cocina_object)
+  end
+
+  def purl_cocina_object_file_map
+    @purl_cocina_object_file_map ||= file_map_for(purl_cocina_object)
+  end
+
+  def purl_cocina_object
+    @purl_cocina_object ||= begin
+      response = connection.get("/#{bare_druid}")
```
Comment (jcoyne) [inline app/services/digital_stacks_differ.rb:39]: Do you need to add `.json` as a suffix here?  Won't it give HTML by default?

File: app/services/digital_stacks_differ.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  def self.call(...)
+    new(...).call
+  end
+
+  def initialize(cocina_object:)
+    @cocina_object = cocina_object
+  end
+
+  # return [Array<String>] filenames of files that should be shelved
```
Comment (jcoyne) [inline app/services/digital_stacks_differ.rb:18]: What about items that should be removed from stacks?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/services/digital_stacks_differ.rb (hunk 2 above)
Comment (justinlittman) [reply app/services/digital_stacks_differ.rb:18]: Stacks needs to figure that out.

File: app/services/digital_stacks_differ.rb (hunk 1 above)
Comment (justinlittman) [reply app/services/digital_stacks_differ.rb:39]: I'd expect so. Fixing.

File: app/services/digital_stacks_differ.rb (hunk 2 above)
Comment (jcoyne) [reply app/services/digital_stacks_differ.rb:18]: Currently we don't have a stacks API, it's just a file system.  I would agree with you, but only after https://github.com/sul-dlss/purl-fetcher/issues/718 is done.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**